File: src/org/simbrain/world/odorworld/sensors/ObjectSensor.java
package org.simbrain.world.odorworld.sensors;

import org.simbrain.util.UserParameter;
import org.simbrain.util.environment.ScalarSmellSource;
import org.simbrain.util.math.SimbrainMath;
import org.simbrain.util.propertyeditor2.EditableObject;
import org.simbrain.workspace.Producible;
import org.simbrain.world.odorworld.entities.OdorWorldEntity;

/**
 * Sensor that reacts when an object of a given type is near it.
 */
public class ObjectSensor extends Sensor {

    /**
     * Current value of the sensor.
     */
    private double value = 0;

    /**
     * The type of the object represented, e.g. Swiss.gif.
     */
    @UserParameter(label = "Object Type",
            description = "What type of object this sensor responds to",
            order = 5)
    private OdorWorldEntity.EntityType objectType = OdorWorldEntity.EntityType.SWISS;

    /**
     * Instantiate an object sensor.
     *
     * @param parent parent entity
     * @param objectType the type (e.g. Swiss.gif)
     */
    public ObjectSensor(OdorWorldEntity parent, OdorWorldEntity.EntityType objectType) {
        super(parent, objectType.toString());
        this.objectType = objectType;
    }

    /**
     * Instantiate an object sensor.
     *
     * @param parent parent entity
     */
    public ObjectSensor(OdorWorldEntity parent) {
        super(parent, "Object Sensor");
    }

    @Override
    public void update() {
        value = 0;
        for (OdorWorldEntity entity : parent.getParentWorld().getEntityList()) {
            ScalarSmellSource smell = entity.getScalarSmell();
            if(entity.getEntityType() == objectType) {
                double distance = SimbrainMath.distance(parent.getCenterLocation(), entity.getCenterLocation());
                value += smell.getValue(distance);
            }
        }
    }

    @Producible(idMethod = "getId", customDescriptionMethod = "getSensorDescription")
    public double getCurrentValue() {
        return value;
    }

    @Override
    public String getTypeDescription() {
        return objectType.toString();
    }

    /**
     * Called by reflection to return a custom description for the {@link org.simbrain.workspace.gui.couplingmanager.AttributePanel.ProducerOrConsumer}
     * corresponding to object sensors.
     */
    public String getSensorDescription() {
        return getParent().getName() + ":" + getTypeDescription() + " sensor";
    }

    @Override
    public EditableObject copy() {
        return new ObjectSensor(parent, objectType);
    }

    @Override
    public String getName() {
        return "ObjectSensor";
    }
}

File: src/main/java/org/simbrain/world/odorworld/sensors/ObjectSensor.kt
package org.simbrain.world.odorworld.sensors

import org.simbrain.util.UserParameter
import org.simbrain.util.decayfunctions.DecayFunction
import org.simbrain.util.decayfunctions.LinearDecayFunction
import org.simbrain.util.math.SimbrainMath
import org.simbrain.util.piccolo.getTilesNear
import org.simbrain.workspace.Producible
import org.simbrain.world.odorworld.entities.EntityType
import org.simbrain.world.odorworld.entities.OdorWorldEntity

/**
 * Sensor that reacts when an object of a given type is near it.
 * <br></br>
 * While the smell framework involves objects emitting smells, object type
 * sensors have a sensitivity, and are more "sensor" or "subject" based than
 * object based.
 * <br></br>
 * The sensor itself is currently fixed at the center of the agent. We may
 * make the location editable at some point, if use-cases emerge.
 */
class ObjectSensor : Sensor, VisualizableEntityAttribute {
    /**
     * Current value of the sensor.
     */
    @get:Producible(customDescriptionMethod = "getAttributeDescription")
    var currentValue = 0.0
        private set

    @UserParameter(
        description = "Maximum value of the sensor when agent is right on top of the associated object type",
        label = "Max Value",
        order = 10
    )
    var baseValue = 1.0
        private set

    /**
     * Decay function
     */
    @UserParameter(label = "Decay Function", isObjectType = true, showDetails = false, order = 15)
    var decayFunction: DecayFunction = LinearDecayFunction(70.0)

    /**
     * The type of the object represented, e.g. Swiss.gif.
     */
    @UserParameter(label = "Object Type", description = "What type of object this sensor responds to", order = 3)
    private var objectType = EntityType.SWISS

    @UserParameter(
        label = "Show dispersion",
        description = "Show dispersion of the sensor",
        useSetter = true,
        order = 4
    )
    @get:JvmName("isShowDispersion")
    var showDispersion = false

    /**
     * Should the sensor node show a label on top.
     */
    @UserParameter(label = "Show Label", description = "Show label on top of the sensor node", order = 5)
    var isShowLabel = false
        private set(value) {
            field = value
            events.firePropertyChanged()
        }

    /**
     * Instantiate an object sensor.
     *
     * @param objectType the type (e.g. Swiss.gif)
     */
    constructor(objectType: EntityType) {
        this.objectType = objectType
    }

    /**
     * Construct a copy of a object sensor.
     *
     * @param objectSensor the object sensor to copy
     */
    constructor(objectSensor: ObjectSensor) : super(objectSensor) {
        setId(objectSensor.id)
        baseValue = objectSensor.baseValue
        decayFunction = objectSensor.decayFunction.copy() as DecayFunction
        objectType = objectSensor.objectType
        isShowLabel = objectSensor.isShowLabel
    }

    /**
     * Default constructor for [org.simbrain.util.propertyeditor.AnnotatedPropertyEditor].
     */
    constructor() : super() {}
    constructor(type: EntityType, radius: Double, angle: Double) {
        objectType = type
        this.radius = radius
        theta = angle
    }

    override fun update(parent: OdorWorldEntity) {
        currentValue = 0.0
        val thing = with(parent.world.tileMap) {
            getTilesNear(parent, decayFunction.dispersion)
                .filter { (_, tile) -> tile.type == "Grass" }
        }

        for (otherEntity in parent.world.entityList) {
            if (otherEntity.entityType == objectType) {
                val scaleFactor = decayFunction.getScalingFactor(
                    SimbrainMath.distance(computeAbsoluteLocation(parent), otherEntity.location)
                )
                currentValue += baseValue * scaleFactor
            }
        }
    }

    override fun copy(): ObjectSensor {
        return ObjectSensor(this)
    }

    override val name: String
        get() = "Object Sensor"

    fun setObjectType(objectType: EntityType) {
        this.objectType = objectType
    }

    override fun getLabel(): String {
        return if (super.getLabel().isEmpty()) {
            directionString + objectType.description + " Detector"
        } else {
            super.getLabel()
        }
    }
}